File: bindings/go/agentgate/models_test.go
package agentgate

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"log/slog"
	"math"
	"strings"
	"testing"
)

func TestIssueRequestValidation(t *testing.T) {
	request, err := NewV1IssueRequest([]byte{0x00, 0xff})
	if err != nil {
		t.Fatalf("NewV1IssueRequest: %v", err)
	}
	if request.Version() != "1.0" || request.AttemptLimit() != AttemptLimitOne {
		t.Fatalf("unexpected request: version=%q limit=%d", request.Version(), request.AttemptLimit())
	}

	invalid := []struct {
		name    string
		version string
		binding []byte
		limit   AttemptLimit
	}{
		{"empty binding", "1.0", nil, AttemptLimitOne},
		{"long binding", "1.0", make([]byte, 257), AttemptLimitOne},
		{"invalid limit", "1.0", []byte("binding"), AttemptLimit(3)},
		{"invalid version UTF-8", string([]byte{0xff}), []byte("binding"), AttemptLimitOne},
	}
	for _, test := range invalid {
		t.Run(test.name, func(t *testing.T) {
			if _, err := NewIssueRequest(test.version, test.binding, test.limit); err == nil {
				t.Fatal("expected validation error")
			}
		})
	}
}

func TestV1IssueRequestDefaultsToOneAttempt(t *testing.T) {
	var constructor func([]byte) (IssueRequest, error) = NewV1IssueRequest
	request, err := constructor([]byte("binding"))
	if err != nil || request.AttemptLimit() != AttemptLimitOne {
		t.Fatalf("unexpected V1 default: %#v, %v", request, err)
	}
}

func TestIssueRequestCopiesBinding(t *testing.T) {
	binding := []byte("binding")
	request, err := NewIssueRequest("1.0", binding, AttemptLimitTwo)
	if err != nil {
		t.Fatalf("NewIssueRequest: %v", err)
	}
	binding[0] = 'X'
	if string(request.binding) != "binding" {
		t.Fatalf("stored binding changed with caller input: %q", request.binding)
	}
}

func TestSecretBearingModelsUseSafeFormattingAndLogging(t *testing.T) {
	issue, err := NewIssueRequest("VERSION_SENTINEL", []byte("BINDING_SENTINEL"), AttemptLimitTwo)
	if err != nil {
		t.Fatalf("NewIssueRequest: %v", err)
	}
	models := []struct {
		name      string
		value     any
		forbidden []string
	}{
		{
			name:  "issue request",
			value: issue,
			forbidden: []string{
				"VERSION_SENTINEL", "BINDING_SENTINEL", fmt.Sprint([]byte("BINDING_SENTINEL")),
			},
		},
		{
			name: "submission",
			value: Submission{
				ChallengeID: "CHALLENGE_SENTINEL", Nonce: "NONCE_SENTINEL", Answer: "ANSWER_SENTINEL",
			},
			forbidden: []string{"NONCE_SENTINEL", "ANSWER_SENTINEL"},
		},
		{
			name: "public challenge",
			value: PublicChallenge{
				ChallengeID: "CHALLENGE_SENTINEL", GeneratorVersion: "1.0", Nonce: "NONCE_SENTINEL",
				IssuedAt: 1, ExpiresAt: 2, Question: "QUESTION_SENTINEL", AnswerEncoding: AnswerEncodingBase64URL,
			},
			forbidden: []string{"NONCE_SENTINEL", "QUESTION_SENTINEL"},
		},
	}

	for _, model := range models {
		t.Run(model.name, func(t *testing.T) {
			outputs := []string{
				fmt.Sprintf("%v", model.value),
				fmt.Sprintf("%+v", model.value),
				fmt.Sprintf("%#v", model.value),
			}

			var standardLog bytes.Buffer
			log.New(&standardLog, "", 0).Print(model.value)
			outputs = append(outputs, standardLog.String())

			var structuredLog bytes.Buffer
			logger := slog.New(slog.NewTextHandler(&structuredLog, nil))
			logger.Info("model", "value", model.value)
			outputs = append(outputs, structuredLog.String())

			for _, output := range outputs {
				for _, sentinel := range model.forbidden {
					if strings.Contains(output, sentinel) {
						t.Errorf("format leaked %q in %q", sentinel, output)
					}
				}
			}
		})
	}
}

func TestSubmissionStrictJSONAndUnicodeRoundTrip(t *testing.T) {
	payload := []byte(`{"challenge_id":"挑战-1","nonce":"bm9uY2U","answer":"YQ"}`)
	got, err := decodeSubmission(payload)
	if err != nil {
		t.Fatalf("decodeSubmission: %v", err)
	}
	if got.ChallengeID != "挑战-1" || got.Nonce != "bm9uY2U" || got.Answer != "YQ" {
		t.Fatalf("unexpected submission: %#v", got)
	}
	encoded, err := encodeSubmission(got)
	if err != nil {
		t.Fatalf("encodeSubmission: %v", err)
	}
	if string(encoded) != string(payload) {
		t.Fatalf("round trip mismatch: %s", encoded)
	}
}

func TestPublicChallengeStrictJSONAndIntegerBounds(t *testing.T) {
	payload := []byte(`{"challenge_id":"id","generator_version":"1.0","nonce":"nonce","issued_at":-9223372036854775808,"expires_at":9223372036854775807,"question":"2 加 2？","answer_encoding":"base64url"}`)
	got, err := decodePublicChallenge(payload)
	if err != nil {
		t.Fatalf("decodePublicChallenge: %v", err)
	}
	if got.IssuedAt != math.MinInt64 || got.ExpiresAt != math.MaxInt64 || got.AnswerEncoding != AnswerEncodingBase64URL {
		t.Fatalf("unexpected challenge: %#v", got)
	}
	encoded, err := encodePublicChallenge(got)
	if err != nil {
		t.Fatalf("encodePublicChallenge: %v", err)
	}
	if string(encoded) != string(payload) {
		t.Fatalf("round trip mismatch: %s", encoded)
	}
}

func TestVerificationOutcomeAcceptedAndRejected(t *testing.T) {
	accepted, err := decodeVerificationOutcome([]byte(`{"status":"accepted"}`))
	if err != nil || accepted.Status != VerificationStatusAccepted || accepted.Reason != "" {
		t.Fatalf("accepted outcome: %#v, %v", accepted, err)
	}

	reasons := []RejectionReason{
		RejectionReasonNotFound,
		RejectionReasonExpired,
		RejectionReasonAlreadyConsumed,
		RejectionReasonBindingMismatch,
		RejectionReasonNonceMismatch,
		RejectionReasonAttemptsExhausted,
	}
	for _, reason := range reasons {
		t.Run(string(reason), func(t *testing.T) {
			payload := []byte(`{"status":"rejected","reason":"` + string(reason) + `"}`)
			outcome, err := decodeVerificationOutcome(payload)
			if err != nil || outcome.Status != VerificationStatusRejected || outcome.Reason != reason {
				t.Fatalf("rejected outcome: %#v, %v", outcome, err)
			}
			encoded, err := encodeVerificationOutcome(outcome)
			if err != nil || string(encoded) != string(payload) {
				t.Fatalf("encode: %s, %v", encoded, err)
			}
		})
	}
}

func TestStrictDecodersRejectMalformedJSON(t *testing.T) {
	validSubmission := `{"challenge_id":"id","nonce":"nonce","answer":"YQ"}`
	cases := map[string]string{
		"unknown field":    `{"challenge_id":"id","nonce":"nonce","answer":"YQ","extra":1}`,
		"wrong case":       `{"Challenge_ID":"id","nonce":"nonce","answer":"YQ"}`,
		"alias collision":  `{"challenge_id":"id","Challenge_ID":"replacement","nonce":"nonce","answer":"YQ"}`,
		"duplicate field":  `{"challenge_id":"id","nonce":"nonce","answer":"YQ","answer":"Yg"}`,
		"trailing value":   validSubmission + ` {}`,
		"missing field":    `{"challenge_id":"id","nonce":"nonce"}`,
		"wrong field type": `{"challenge_id":"id","nonce":3,"answer":"YQ"}`,
		"not an object":    `[]`,
		"invalid UTF-8":    "{\"challenge_id\":\"\xff\",\"nonce\":\"nonce\",\"answer\":\"YQ\"}",
	}
	for name, payload := range cases {
		t.Run(name, func(t *testing.T) {
			if _, err := decodeSubmission([]byte(payload)); !errors.Is(err, errInvalidJSON) {
				t.Fatalf("expected stable invalid JSON error, got %v", err)
			}
		})
	}
}

func TestStrictDecodersRejectWrongCaseKeys(t *testing.T) {
	challenge := []byte(`{"challenge_id":"id","generator_version":"1.0","nonce":"nonce","issued_at":1,"expires_at":2,"question":"q","Answer_Encoding":"base64url"}`)
	if _, err := decodePublicChallenge(challenge); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("accepted wrong-case challenge key: %v", err)
	}

	outcome := []byte(`{"STATUS":"accepted"}`)
	if _, err := decodeVerificationOutcome(outcome); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("accepted wrong-case outcome key: %v", err)
	}
}

func TestStrictDecoderRejectsUnpairedUnicodeSurrogates(t *testing.T) {
	for _, escaped := range []string{`\ud800`, `\udbff`, `\udc00`, `\udfff`} {
		payload := []byte(`{"challenge_id":"` + escaped + `","nonce":"nonce","answer":"YQ"}`)
		if _, err := decodeSubmission(payload); !errors.Is(err, errInvalidJSON) {
			t.Errorf("accepted unpaired surrogate %s: %v", escaped, err)
		}
	}
	got, err := decodeSubmission([]byte(`{"challenge_id":"\ud83d\ude00","nonce":"nonce","answer":"YQ"}`))
	if err != nil || got.ChallengeID != "😀" {
		t.Fatalf("valid surrogate pair was not preserved: %#v, %v", got, err)
	}
}

func TestStrictDecoderHandlesEscapedBackslashesBeforeSurrogates(t *testing.T) {
	got, err := decodeSubmission([]byte(`{"challenge_id":"\\ud800","nonce":"nonce","answer":"YQ"}`))
	if err != nil || got.ChallengeID != `\ud800` {
		t.Fatalf("escaped backslash was treated as a surrogate: %#v, %v", got, err)
	}
}

func TestDuplicateKeyPreScanUsesDecodedKeysAndTraversesArrays(t *testing.T) {
	var escaped struct {
		Answer string `json:"answer"`
	}
	if err := strictDecodeObject([]byte(`{"answer":"YQ","\u0061nswer":"Yg"}`), &escaped); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("expected semantic duplicate rejection, got %v", err)
	}

	var nested struct {
		Items []map[string]any `json:"items"`
	}
	if err := strictDecodeObject([]byte(`{"items":[{"key":1,"key":2}]}`), &nested); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("expected duplicate rejection inside array, got %v", err)
	}
}

func TestDuplicateKeyPreScanRejectsNestedObjects(t *testing.T) {
	var target struct {
		Nested map[string]any `json:"nested"`
	}
	if err := strictDecodeObject([]byte(`{"nested":{"key":1,"key":2}}`), &target); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("expected nested duplicate rejection, got %v", err)
	}
}

func TestPublicChallengeRejectsInvalidFields(t *testing.T) {
	base := `{"challenge_id":"id","generator_version":"1.0","nonce":"nonce","issued_at":1,"expires_at":2,"question":"q","answer_encoding":"base64url"}`
	cases := map[string]string{
		"missing time":       strings.Replace(base, `"issued_at":1,`, "", 1),
		"null time":          strings.Replace(base, `"issued_at":1`, `"issued_at":null`, 1),
		"fractional time":    strings.Replace(base, `"issued_at":1`, `"issued_at":1.5`, 1),
		"time overflow":      strings.Replace(base, `"issued_at":1`, `"issued_at":9223372036854775808`, 1),
		"unknown encoding":   strings.Replace(base, `"base64url"`, `"BASE64URL"`, 1),
		"encoding is null":   strings.Replace(base, `"base64url"`, `null`, 1),
		"duplicate encoding": strings.Replace(base, `"answer_encoding":"base64url"`, `"answer_encoding":"base64url","answer_encoding":"base64url"`, 1),
	}
	for name, payload := range cases {
		t.Run(name, func(t *testing.T) {
			if _, err := decodePublicChallenge([]byte(payload)); !errors.Is(err, errInvalidJSON) {
				t.Fatalf("expected invalid JSON error, got %v", err)
			}
		})
	}
}

func TestVerificationOutcomeRejectsInvalidCombinations(t *testing.T) {
	cases := []string{
		`{"status":"accepted","reason":"expired"}`,
		`{"status":"rejected"}`,
		`{"status":"rejected","reason":null}`,
		`{"status":"unknown"}`,
		`{"status":"rejected","reason":"unknown"}`,
		`{"status":"rejected","reason":"expired","extra":true}`,
	}
	for _, payload := range cases {
		if _, err := decodeVerificationOutcome([]byte(payload)); !errors.Is(err, errInvalidJSON) {
			t.Errorf("%s: expected invalid JSON error, got %v", payload, err)
		}
	}

	invalidModels := []VerificationOutcome{
		{Status: VerificationStatusAccepted, Reason: RejectionReasonExpired},
		{Status: VerificationStatusRejected},
		{Status: VerificationStatus("unknown")},
		{Status: VerificationStatusRejected, Reason: RejectionReason("unknown")},
	}
	for _, outcome := range invalidModels {
		if _, err := encodeVerificationOutcome(outcome); !errors.Is(err, errInvalidJSON) {
			t.Errorf("%#v: expected invalid JSON error, got %v", outcome, err)
		}
	}
}

func TestEncodeRejectsInvalidStringsAndEnums(t *testing.T) {
	bad := string([]byte{0xff})
	if _, err := encodeSubmission(Submission{ChallengeID: bad, Nonce: "nonce", Answer: "YQ"}); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("submission invalid UTF-8: %v", err)
	}
	challenge := PublicChallenge{
		ChallengeID: "id", GeneratorVersion: "1.0", Nonce: "nonce",
		IssuedAt: 1, ExpiresAt: 2, Question: "q", AnswerEncoding: AnswerEncoding("unknown"),
	}
	if _, err := encodePublicChallenge(challenge); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("challenge invalid enum: %v", err)
	}
}

func TestJSONMethodsUseStrictCodec(t *testing.T) {
	var submission Submission
	if err := json.Unmarshal([]byte(`{"challenge_id":"id","nonce":"nonce","answer":"YQ","extra":true}`), &submission); !errors.Is(err, errInvalidJSON) {
		t.Fatalf("UnmarshalJSON was not strict: %v", err)
	}
	payload, err := json.Marshal(Submission{ChallengeID: "挑战", Nonce: "n", Answer: "a"})
	if err != nil || string(payload) != `{"challenge_id":"挑战","nonce":"n","answer":"a"}` {
		t.Fatalf("MarshalJSON mismatch: %s, %v", payload, err)
	}
}

func TestAgentGateErrorMapsStableCodesAndFailsClosed(t *testing.T) {
	cases := map[int32]string{
		1: "invalid_configuration", 2: "generation_failed", 3: "invalid_challenge_material",
		4: "invalid_answer_encoding", 5: "answer_mismatch", 6: "unsupported_generator_version",
		7: "internal_error", 100: "invalid_argument", 101: "callback_failed", 102: "panic_caught",
	}
	if err := errorForStatus(0); err != nil {
		t.Fatalf("OK returned error: %v", err)
	}
	for status, code := range cases {
		err := errorForStatus(status)
		var agentGateErr *AgentGateError
		if !errors.As(err, &agentGateErr) {
			t.Fatalf("status %d: wrong error type %T", status, err)
		}
		if agentGateErr.Code() != code || agentGateErr.Error() != code {
			t.Fatalf("status %d: got code=%q error=%q", status, agentGateErr.Code(), agentGateErr.Error())
		}
	}
	for _, status := range []int32{-1, 8, 99, 103, math.MaxInt32} {
		err := errorForStatus(status)
		var agentGateErr *AgentGateError
		if !errors.As(err, &agentGateErr) || agentGateErr.Code() != "internal_error" || agentGateErr.Error() != "internal_error" {
			t.Fatalf("unknown status %d did not fail closed: %v", status, err)
		}
	}
}

File: bindings/node/test/models.test.js
import assert from 'node:assert/strict';
import { inspect } from 'node:util';
import test from 'node:test';

import { AgentGateError, errorForStatus } from '../lib/errors.js';
import {
  decodePublicChallenge,
  decodeSubmission,
  decodeVerificationOutcome,
  encodePublicChallenge,
  encodeSubmission,
  encodeVerificationOutcome,
} from '../lib/json.js';
import {
  AnswerEncoding,
  AttemptLimit,
  IssueRequest,
  PublicChallenge,
  RejectionReason,
  Submission,
  VerificationOutcome,
  VerificationStatus,
  newV1IssueRequest,
} from '../lib/models.js';

const utf8 = (value) => Buffer.from(value, 'utf8');

function assertInvalid(action) {
  assert.throws(action, (error) => {
    assert.ok(error instanceof AgentGateError);
    assert.equal(error.code, 'invalid_argument');
    assert.equal(error.message, 'invalid_argument');
    assert.equal(String(error), 'invalid_argument');
    return true;
  });
}

test('v1 issue request has fixed defaults and defensive binding copies', () => {
  const source = Uint8Array.from([1, 2, 3]);
  const request = newV1IssueRequest(source);
  source[0] = 9;
  const returned = request.binding;
  returned[1] = 9;

  assert.equal(request.version, '1.0');
  assert.equal(request.attemptLimit, AttemptLimit.ONE);
  assert.deepEqual(request.binding, Uint8Array.from([1, 2, 3]));
  assert.deepEqual(Object.keys(request), ['version', 'binding', 'attemptLimit']);
  assert.ok(Object.isFrozen(request));
  assert.ok(Object.isFrozen(AttemptLimit));
  assert.deepEqual(AttemptLimit, { ONE: 1, TWO: 2 });
});

test('issue requests reject invalid values and non-exact object shapes', () => {
  const valid = { version: '1.0', binding: Uint8Array.of(1), attemptLimit: 1 };
  for (const value of [
    null,
    [],
    Object.create(null),
    { ...valid, extra: true },
    { version: '1.0', binding: Uint8Array.of(1) },
    { ...valid, version: null },
    { ...valid, version: 'bad\ud800' },
    { ...valid, binding: null },
    { ...valid, binding: Uint8Array.of() },
    { ...valid, binding: new Uint8Array(257) },
    { ...valid, attemptLimit: 3 },
  ]) {
    assertInvalid(() => new IssueRequest(value));
  }
  const inherited = Object.create(valid);
  Object.assign(inherited, valid);
  assertInvalid(() => new IssueRequest(inherited));
  const symbol = { ...valid, [Symbol('hidden')]: true };
  assertInvalid(() => new IssueRequest(symbol));
  const accessor = { ...valid };
  Object.defineProperty(accessor, 'version', { enumerable: true, get: () => '1.0' });
  assertInvalid(() => new IssueRequest(accessor));
  assertInvalid(() => new IssueRequest(valid, 'ignored'));
  assertInvalid(() => newV1IssueRequest(Uint8Array.of(1), 'ignored'));
});

test('public models require exact plain-object shapes and freeze outputs', () => {
  const challengeValues = {
    challengeId: 'challenge', generatorVersion: '1.0', nonce: 'nonce',
    issuedAt: 1, expiresAt: 2, question: 'question', answerEncoding: 'base64url',
  };
  const submissionValues = { challengeId: 'challenge', nonce: 'nonce', answer: 'answer' };
  const challenge = new PublicChallenge(challengeValues);
  const submission = new Submission(submissionValues);

  assert.deepEqual(Object.keys(challenge), Object.keys(challengeValues));
  assert.deepEqual(Object.keys(submission), Object.keys(submissionValues));
  assert.ok(Object.isFrozen(challenge));
  assert.ok(Object.isFrozen(submission));
  assert.throws(() => { challenge.question = 'changed'; }, TypeError);
  assert.equal(challenge.question, 'question');

  for (const changed of [
    { ...challengeValues, issuedAt: 1.5 },
    { ...challengeValues, expiresAt: Number.MAX_SAFE_INTEGER + 1 },
    { ...challengeValues, answerEncoding: 'BASE64URL' },
    { ...challengeValues, question: 'bad\udfff' },
    { ...challengeValues, Extra: true },
  ]) assertInvalid(() => new PublicChallenge(changed));
  assertInvalid(() => new Submission({ ...submissionValues, answer: 1 }));
  assertInvalid(() => new Submission({ ...submissionValues, Answer: 'answer' }));
  assertInvalid(() => new Submission(Object.assign(Object.create({ polluted: true }), submissionValues)));
  assertInvalid(() => new PublicChallenge(challengeValues, 'ignored'));
  assertInvalid(() => new Submission(submissionValues, 'ignored'));
});

test('typed model construction copies explicit fields and rejects prototype tricks', () => {
  const values = { challengeId: 'c', nonce: 'n', answer: 'a' };
  const model = new Submission(values);
  values.answer = 'changed';
  values.extra = 'later';
  assert.equal(model.answer, 'a');
  assert.deepEqual(Object.keys(model), ['challengeId', 'nonce', 'answer']);

  const nonEnumerable = { ...values };
  delete nonEnumerable.extra;
  Object.defineProperty(nonEnumerable, 'hidden', { value: true });
  assertInvalid(() => new Submission(nonEnumerable));
});

test('model constructors reject malicious subclasses before they can leak secrets', () => {
  const sentinel = 'SUBCLASS_SECRET_SENTINEL';
  class LeakyIssueRequest extends IssueRequest {
    toJSON() { return sentinel; }
    [inspect.custom]() { return sentinel; }
  }
  class LeakyPublicChallenge extends PublicChallenge {
    toJSON() { return sentinel; }
    [inspect.custom]() { return sentinel; }
  }
  class LeakySubmission extends Submission {
    toJSON() { return sentinel; }
    [inspect.custom]() { return sentinel; }
  }
  class LeakyVerificationOutcome extends VerificationOutcome {
    toJSON() { return sentinel; }
    [inspect.custom]() { return sentinel; }
  }

  const cases = [
    () => new LeakyIssueRequest({
      version: '1.0', binding: Uint8Array.of(1), attemptLimit: AttemptLimit.ONE,
    }),
    () => new LeakyPublicChallenge({
      challengeId: 'c', generatorVersion: '1.0', nonce: sentinel,
      issuedAt: 1, expiresAt: 2, question: sentinel,
      answerEncoding: AnswerEncoding.BASE64URL,
    }),
    () => new LeakySubmission({ challengeId: 'c', nonce: sentinel, answer: sentinel }),
    () => new LeakyVerificationOutcome(Symbol(sentinel), 'accepted', null),
  ];
  for (const construct of cases) {
    let caught;
    try { construct(); } catch (error) { caught = error; }
    assert.ok(caught instanceof AgentGateError);
    assert.equal(caught.code, 'invalid_argument');
    for (const rendered of [String(caught), inspect(caught), JSON.stringify(caught)]) {
      assert.equal(rendered.includes(sentinel), false);
    }
  }
});

test('issue requests require an exact genuine Uint8Array and normalize intrinsic failures', () => {
  class DerivedBytes extends Uint8Array {}
  const customPrototype = Object.create(Uint8Array.prototype);
  const customView = Uint8Array.of(1);
  Object.setPrototypeOf(customView, customPrototype);
  const proxy = new Proxy(Uint8Array.of(1), {});
  const detached = Uint8Array.of(1);
  structuredClone(detached.buffer, { transfer: [detached.buffer] });

  for (const binding of [
    Object.create(Uint8Array.prototype),
    new DerivedBytes([1]),
    Buffer.from([1]),
    customView,
    proxy,
    detached,
  ]) {
    assertInvalid(() => newV1IssueRequest(binding));
  }
});

test('model string, inspection, and implicit JSON forms redact secrets', () => {
  const sentinels = ['BINDING_SENTINEL', 'NONCE_SENTINEL', 'ANSWER_SENTINEL', 'QUESTION_SENTINEL'];
  const models = [
    newV1IssueRequest(Uint8Array.from(utf8(sentinels[0]))),
    new PublicChallenge({
      challengeId: 'challenge', generatorVersion: '1.0', nonce: sentinels[1],
      issuedAt: 1, expiresAt: 2, question: sentinels[3], answerEncoding: AnswerEncoding.BASE64URL,
    }),
    new Submission({ challengeId: 'challenge', nonce: sentinels[1], answer: sentinels[2] }),
  ];
  for (const model of models) {
    for (const rendered of [String(model), inspect(model), JSON.stringify(model)]) {
      for (const sentinel of sentinels) assert.equal(rendered.includes(sentinel), false);
    }
  }
});

test('submission round-trips with exact deterministic bytes and non-BMP Unicode', () => {
  const expected = new Submission({ challengeId: '雪🚀', nonce: 'nonce', answer: 'answer' });
  const encoded = encodeSubmission(expected);
  assert.ok(encoded instanceof Uint8Array);
  assert.equal(Buffer.from(encoded).toString('utf8'),
    '{"challenge_id":"雪🚀","nonce":"nonce","answer":"answer"}');
  assert.deepEqual(decodeSubmission(encoded), expected);
});

test('public challenge round-trips with exact safe integers and encoding', () => {
  const expected = new PublicChallenge({
    challengeId: 'challenge', generatorVersion: '1.0', nonce: 'nonce',
    issuedAt: Number.MIN_SAFE_INTEGER, expiresAt: Number.MAX_SAFE_INTEGER,
    question: '雪🚀', answerEncoding: AnswerEncoding.BASE64URL,
  });
  const encoded = encodePublicChallenge(expected);
  assert.equal(Buffer.from(encoded).toString('utf8'),
    '{"challenge_id":"challenge","generator_version":"1.0","nonce":"nonce",' +
      '"issued_at":-9007199254740991,"expires_at":9007199254740991,' +
      '"question":"雪🚀","answer_encoding":"base64url"}');
  assert.deepEqual(decodePublicChallenge(encoded), expected);
});

test('verification outcomes round-trip with exact closed shapes', () => {
  const accepted = VerificationOutcome.accepted();
  assert.equal(Buffer.from(encodeVerificationOutcome(accepted)).toString(), '{"status":"accepted"}');
  assert.deepEqual(decodeVerificationOutcome(encodeVerificationOutcome(accepted)), accepted);

  const reasons = Object.values(RejectionReason);
  assert.equal(reasons.length, 6);
  for (const reason of reasons) {
    const rejected = VerificationOutcome.rejected(reason);
    const exact = `{"status":"rejected","reason":"${reason}"}`;
    assert.equal(Buffer.from(encodeVerificationOutcome(rejected)).toString(), exact);
    assert.deepEqual(decodeVerificationOutcome(utf8(exact)), rejected);
    assert.ok(Object.isFrozen(rejected));
  }
  assert.deepEqual(VerificationStatus, { ACCEPTED: 'accepted', REJECTED: 'rejected' });
  assertInvalid(() => VerificationOutcome.accepted('ignored'));
  assertInvalid(() => VerificationOutcome.rejected(RejectionReason.EXPIRED, 'ignored'));
});

test('duplicate keys are rejected at every nesting depth and after escape decoding', () => {
  for (const json of [
    '{"status":"accepted","status":"accepted"}',
    '{"challenge_id":"c","nonce":"n","answer":"a","extra":{"same":1,"same":2}}',
    '{"challenge_id":"c","nonce":"n","answer":"a","extra":[{"same":1,"s\\u0061me":2}]}',
    '{"challenge_id":"c","nonce":"n","answer":"a","extra":{"\\\"":1,"\\u0022":2}}',
  ]) assertInvalid(() => decodeVerificationOutcome(utf8(json)));
});

test('strict scanner rejects malformed syntax, trailing roots, and invalid UTF-8', () => {
  for (const bytes of [
    utf8('{"status":"accepted"} {"status":"accepted"}'),
    utf8('{"status":"accepted",}'),
    utf8('{"status":"accepted"'),
    utf8('{status:"accepted"}'),
    utf8('{"status":tru}'),
    utf8('{"status":"accepted"} garbage'),
    Uint8Array.from([0x7b, 0x22, 0x78, 0x22, 0x3a, 0x22, 0xc3, 0x28, 0x22, 0x7d]),
  ]) assertInvalid(() => decodeVerificationOutcome(bytes));
});

test('strict scanner accepts JSON whitespace, escapes, arrays, and valid raw Unicode', () => {
  assert.deepEqual(
    decodeSubmission(utf8(' \n\t {"challenge_id":"雪🚀","nonce":"n","answer":"a"}\r ')),
    new Submission({ challengeId: '雪🚀', nonce: 'n', answer: 'a' }),
  );
  assertInvalid(() => decodeSubmission(utf8(
    '{"challenge_id":"c","nonce":"n","answer":"a","extra":[1,true,null,{"x":"y"}]}')));
});

test('lone and unpaired surrogates are rejected in values and keys', () => {
  for (const json of [
    '{"challenge_id":"\\uD800","nonce":"n","answer":"a"}',
    '{"challenge_id":"\\uDC00","nonce":"n","answer":"a"}',
    '{"challenge_id":"\\uD800x","nonce":"n","answer":"a"}',
    '{"\\uD800":"x","nonce":"n","answer":"a"}',
  ]) assertInvalid(() => decodeSubmission(utf8(json)));
  assertInvalid(() => encodeSubmission({ challengeId: 'bad\ud800', nonce: 'n', answer: 'a' }));
});

test('decoders reject missing, extra, wrong-case, wrong-token and non-object roots', () => {
  for (const json of [
    '{"challenge_id":"c","nonce":"n"}',
    '{"challenge_id":"c","nonce":"n","answer":"a","extra":true}',
    '{"Challenge_id":"c","nonce":"n","answer":"a"}',
    '{"challenge_id":1,"nonce":"n","answer":"a"}',
    '{"challenge_id":null,"nonce":"n","answer":"a"}',
    '[]', 'null', '"object"',
  ]) assertInvalid(() => decodeSubmission(utf8(json)));
});

test('public challenge decoder rejects floats, exponents, unsafe integers and enums', () => {
  const make = (issuedAt, encoding = 'base64url') => utf8(
    `{"challenge_id":"c","generator_version":"1.0","nonce":"n",` +
    `"issued_at":${issuedAt},"expires_at":2,"question":"q","answer_encoding":"${encoding}"}`);
  for (const input of [make('1.0'), make('1e0'), make('9007199254740992'), make('1', 'BASE64URL')]) {
    assertInvalid(() => decodePublicChallenge(input));
  }
});

test('outcome decoder rejects unknown status/reason and inconsistent shapes', () => {
  for (const json of [
    '{"status":"unknown"}',
    '{"status":"rejected","reason":"unknown"}',
    '{"status":"rejected"}',
    '{"status":"accepted","reason":"expired"}',
    '{"Status":"accepted"}',
  ]) assertInvalid(() => decodeVerificationOutcome(utf8(json)));
});

test('encoders accept only exact model instances and never pass through extra fields', () => {
  const submission = new Submission({ challengeId: 'c', nonce: 'n', answer: 'a' });
  assert.equal(Buffer.from(encodeSubmission(submission)).toString(),
    '{"challenge_id":"c","nonce":"n","answer":"a"}');
  assertInvalid(() => encodeSubmission({ challengeId: 'c', nonce: 'n', answer: 'a' }));
  assertInvalid(() => encodePublicChallenge({}));
  assertInvalid(() => encodeVerificationOutcome({ status: 'accepted' }));
});

test('encoders reject objects forged from public model prototypes', () => {
  const forgedSubmission = Object.assign(Object.create(Submission.prototype), {
    challengeId: 'c', nonce: 'n', answer: 'a',
  });
  const forgedChallenge = Object.assign(Object.create(PublicChallenge.prototype), {
    challengeId: 'c', generatorVersion: '1.0', nonce: 'n', issuedAt: 1, expiresAt: 2,
    question: 'q', answerEncoding: AnswerEncoding.BASE64URL,
  });
  const forgedOutcome = Object.assign(Object.create(VerificationOutcome.prototype), {
    status: VerificationStatus.ACCEPTED, reason: null,
  });
  assertInvalid(() => encodeSubmission(forgedSubmission));
  assertInvalid(() => encodePublicChallenge(forgedChallenge));
  assertInvalid(() => encodeVerificationOutcome(forgedOutcome));
});

test('native status mapping is exact and unknown values fail closed', () => {
  assert.equal(errorForStatus(0), null);
  const statuses = [1, 2, 3, 4, 5, 6, 7, 100, 101, 102];
  const codes = [
    'invalid_configuration', 'generation_failed', 'invalid_challenge_material',
    'invalid_answer_encoding', 'answer_mismatch', 'unsupported_generator_version',
    'internal_error', 'invalid_argument', 'callback_failed', 'panic_caught',
  ];
  assert.deepEqual(statuses.map((status) => errorForStatus(status).code), codes);
  assert.equal(errorForStatus(-1).code, 'internal_error');
  assert.equal(errorForStatus(999).code, 'internal_error');
});

test('errors expose and render only their stable code', () => {
  const error = errorForStatus(5, new Error('NATIVE_SECRET'));
  assert.ok(error instanceof Error);
  assert.ok(error instanceof AgentGateError);
  assert.deepEqual(Object.keys(error), ['code']);
  assert.equal(error.name, 'AgentGateError');
  assert.equal(error.message, 'answer_mismatch');
  assert.equal(error.stack, undefined);
  assert.equal(error.cause, undefined);
  for (const rendered of [String(error), inspect(error), JSON.stringify(error)]) {
    assert.equal(rendered.includes('answer_mismatch'), true);
    assert.equal(rendered.includes('NATIVE_SECRET'), false);
  }
});
